File: Number_of_Dice_Rolls_with_target_Sum.py
# Title: 1155. Number of Dice Rolls With Target Sum
# Difficulty: Medium
# Problem: https://leetcode.com/problems/number-of-dice-rolls-with-target-sum/description

class Solution:
    def numRollsToTarget(self, n: int, k: int, target: int) -> int:
        # use moization using cache function
        @cache
        def find(n: int, target: int) -> int:
            if n < 0 or target < 0: return 0 # zeros ways
            if target == 0 and n == 0:
                return 1 # one way
            total = 0
            for i in range(1, k+1):
                total += find(n-1, target-i) # add all the number ways
            return total % (10 ** 9 + 7)
        return find(n, target)

File: Number_of_Dice_Rolls_with_Target_Sum.cpp
// Title: 1155. Number of Dice Rolls With Target Sum
// Difficulty: Medium
// Problem: https://leetcode.com/problems/number-of-dice-rolls-with-target-sum/description

class Solution {
public:
    int numRollsToTarget(int n, int k, int target) {
        int dp[30][1001] = {0};
        int i, j;
        int mod = 1e9 + 7;
        for (j = 1; j <= k && j <= target; j++) {
            dp[0][j] = 1;
        }
        for (i = 1; i < n; i++) {
            for (j = 1; j <= target; j++) {
                dp[i][j] = (dp[i - 1][j - 1] + dp[i][j - 1]) % mod;
                if (j > k) {
                    dp[i][j] = (dp[i][j] - dp[i - 1][j - 1 - k] + mod) % mod;
                }
            }
        }
        return dp[n - 1][target];
    }
};
